### PR TITLE
feat: integrate slack app

### LINKED ISSUE #1050: [Feature Request] Integrate slack app
### Required prerequisites

- [X] I have searched the [Issue Tracker](https://github.com/camel-ai/camel/issues) and [Discussions](https://github.com/camel-ai/camel/discussions) that this hasn't already been reported. (+1 or comment there if it has.)
- [ ] Consider asking first in a [Discussion](https://github.com/camel-ai/camel/discussions/new).

### Motivation

Integrate slack app

### Solution

_No response_

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by koch3092**
## Description

Describe your changes in detail.

## Motivation and Context

close #1050 

- [ ] I have raised an issue to propose this change ([required](https://github.com/camel-ai/camel/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of example)

## Implemented Tasks

- [ ] Subtask 1
- [ ] Subtask 2
- [ ] Subtask 3

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: camel/bots/slack/slack_app.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+            say (AsyncSay): A function to send a response back to the channel.
+        """
+        event_profile = SlackEventProfile(**event)
+        event_body = SlackEventBody(**body)
+
+        logger.info("on_message, context: {}".format(context))
+        logger.info("on_message, client: {}".format(client))
+        logger.info("on_message, event_profile: {}".format(event_profile))
+        logger.info("on_message, event_body: {}".format(event_body))
+        logger.info("on_message, say: {}".format(say))
```
Comment (Wendong-Fan) [inline camel/bots/slack/slack_app.py:226]: use `f`, same as logger below

```suggestion
        logger.info(f"on_message, context: {context}")
        logger.info(f"on_message, client: {client}")
        logger.info(f"on_message, event_profile: {event_profile}")
        logger.info(f"on_message, event_body: {event_body}")
        logger.info(f"on_message, say: {say}")
```

File: camel/bots/slack/slack_app.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+        This method is triggered when someone mentions the app in Slack.
+
+        Args:
+            context (AsyncBoltContext): The Slack Bolt context for the event.
+            client (AsyncWebClient): The Slack Web API client.
+            event (dict): The event data for the app mention.
+            body (dict): The full request body from Slack.
+            say (AsyncSay): A function to send a response back to the channel.
+        """
+        event_profile = SlackEventProfile(**event)
```
Comment (Wendong-Fan) [inline camel/bots/slack/slack_app.py:219]: should we also add ` await context.ack()`?

File: camel/bots/slack/slack_app.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+        Args:
+            request (Request): A Starlette request object representing the
+                incoming request.
+
+        Returns:
+            The response generated by the Slack Bolt handler.
+        """
+        return await self._handler.handle(request)
+
+    async def app_mention(
```
Comment (Wendong-Fan) [inline camel/bots/slack/slack_app.py:200]: add error handing with `try expect`, same as other methods could raise potential error

File: camel/bots/slack/slack_app.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        token: Optional[str] = None,
+        scopes: Optional[str] = None,
+        signing_secret: Optional[str] = None,
+        client_id: Optional[str] = None,
+        client_secret: Optional[str] = None,
+        redirect_uri_path: str = "/slack/oauth_redirect",
+        installation_store: Optional[AsyncInstallationStore] = None,
+    ) -> None:
+        """Initializes the SlackApp instance by setting up the Slack Bolt app
+        and configuring event handlers and OAuth settings.
```
Comment (Wendong-Fan) [inline camel/bots/slack/slack_app.py:71]: docstring format, same as doc below

```suggestion
        r"""Initializes the SlackApp instance by setting up the Slack Bolt app
        and configuring event handlers and OAuth settings.
```

File: camel/bots/slack/slack_app.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+            The response generated by the Slack Bolt handler.
+        """
+        return await self._handler.handle(request)
+
+    async def app_mention(
+        self,
+        context: "AsyncBoltContext",
+        client: "AsyncWebClient",
+        event: dict,
+        body: dict,
```
Comment (Wendong-Fan) [inline camel/bots/slack/slack_app.py:205]: could we make the type hint more specific? like `Dict[str]`

File: camel/bots/slack/slack_app.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+        Args:
+            port (int): The port on which the server should run (default is
+                3000).
+            path (str): The endpoint path for receiving Slack events (default
+                is "/slack/events").
+            host (Optional[str]): The hostname to bind the server (default is
+                None).
+        """
+        thread = threading.Thread(target=self.run, args=(port, path, host))
+        thread.start()
```
Comment (Wendong-Fan) [inline camel/bots/slack/slack_app.py:184]: should we add `thread.daemon = True ` to make the thread run in the background?

File: camel/bots/slack/models.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+from typing import Optional
+
+from pydantic import BaseModel
+
+
+class SlackAuthProfile(BaseModel):
+    """Represents the authorization profile within a Slack event.
```
Comment (Wendong-Fan) [inline camel/bots/slack/models.py:20]: docstring format, same as below

```suggestion
    r"""Represents the authorization profile within a Slack event.
```

File: examples/bots/slack_bot.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+import asyncio
+import queue
+from typing import Optional
+
+from slack_bolt.context.async_context import AsyncBoltContext
+from slack_bolt.context.say.async_say import AsyncSay
+from slack_sdk.web.async_client import AsyncWebClient
+
+from camel.bots import SlackApp, SlackEventBody
+from examples.bots.agent import Agent
```
Comment (Wendong-Fan) [inline examples/bots/slack_bot.py:23]: for the example, I feel we need to make related code into one python file, if there's some abstraction need to be further added, it shouldn't be added into example folder

**Review by Wendong-Fan — COMMENTED** [review]
Thanks @koch3092 , left some comments

**Review by Wendong-Fan — COMMENTED** [review]
please remember to run `poetry lock` to update the lock file after you update the poetry toml file

File: camel/bots/slack/slack_app.py (hunk 6 above)
Comment (koch3092) [reply camel/bots/slack/slack_app.py:184]: I adjusted the thread calling relationship and this method will be deleted

File: examples/bots/slack_bot.py (hunk 8 above)
Comment (koch3092) [reply examples/bots/slack_bot.py:23]: `slack_bot` is an application of `slack_app`. Here we only introduce one usage, so we put it in the examples, and I think users may not necessarily use all platforms, so I think it makes sense to list them separately for them to choose.

Comment (koch3092) [reply examples/bots/slack_bot.py:23]: But for the `camel.bots` module, I have thought about abstracting them, but maybe after integrating more platforms and abstracting away the common process

Comment (koch3092) [reply examples/bots/slack_bot.py:23]: btw, what do you think about renaming `camel.bots` to `camel.adapters`, or something else? I think `bots` are also okay, but it always feels a bit strange that there are a bunch of platform apps in them. (although telegram is still a bot)

File: camel/bots/slack/slack_app.py (hunk 3 above)
Comment (koch3092) [reply camel/bots/slack/slack_app.py:200]: I didn't even find any exceptions raised in the official relevant code. I think this place can be left to the user to handle. Or if there are some problems during specific application, I will add them.

File: examples/bots/slack_bot.py (hunk 8 above)
Comment (Wendong-Fan) [reply examples/bots/slack_bot.py:25]: let's keep the naming as `bots` as that's our target to integrate those apps

File: camel/bots/slack/slack_app.py [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+                signing_secret=self.signing_secret,
+                installation_store=installation_store,
+                token=self.token,
+            )
+
+        self._handler = AsyncSlackRequestHandler(self._app)
+        self.setup_handlers()
+
+    def setup_handlers(self) -> None:
+        """Sets up the event handlers for Slack events, such as `app_mention`
```
Comment (Wendong-Fan) [inline camel/bots/slack/slack_app.py:138]:
```suggestion
        r"""Sets up the event handlers for Slack events, such as `app_mention`
```

File: pyproject.toml [hunk 10]
```diff
@@ -114,6 +114,7 @@ cohere = { version = "^4.56", optional = true }
 pytest = { version = "^7", optional = true }
 pytest-asyncio = { version = "^0.23.0", optional = true }
 mock = { version = "^5", optional = true }
+slack-bolt = "^1.20.1"
```
Comment (Wendong-Fan) [inline pyproject.toml:117]: please set this dependency as optional, you can refer to how other tools are added

**Comment by Wendong-Fan** [comment]
please refer to https://github.com/camel-ai/camel/pull/1073/files @koch3092 
1. update the dependency in toml file and lock file
2. docstring set default value use :obj:

**Review by Wendong-Fan — APPROVED** [review]

**Pull request merged**